pkg/search/postgres/store.go

```diff
@@ -0,0 +1,261 @@
+package postgres
+
+import (
+	"context"
+	"time"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/jackc/pgx/v4"
+	"github.com/pkg/errors"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/pkg/auth/permissions"
+	ops "github.com/stackrox/rox/pkg/metrics"
+	"github.com/stackrox/rox/pkg/postgres"
+	"github.com/stackrox/rox/pkg/postgres/pgutils"
+	"github.com/stackrox/rox/pkg/postgres/walker"
+	"github.com/stackrox/rox/pkg/search"
+)
+
+const (
+	cursorBatchSize = 50
+	deleteBatchSize = 5000
+)
+
+// GenericSingleIDStore implements subset of Store interface for resources with single ID.
+type GenericSingleIDStore[T any, PT singleID[T]] struct {
+	db                               postgres.DB
+	typ                              string
+	targetResource                   permissions.ResourceMetadata
+	schema                           *walker.Schema
+	setPostgresOperationDurationTime func(start time.Time, op ops.Op, t string)
+}
+
+type singleID[T any] interface {
+	proto.Unmarshaler
+	*T
+}
+
+// NewGenericSingleIDStore returns new subStore implementation for given resource.
+// subStore implements subset of Store operations.
+func NewGenericSingleIDStore[T any, PT singleID[T]](db postgres.DB, typ string, targetResource permissions.ResourceMetadata, schema *walker.Schema, setPostgresOperationDurationTime func(start time.Time, op ops.Op, t string)) *GenericSingleIDStore[T, PT] {
+	return &GenericSingleIDStore[T, PT]{
+		db:                               db,
+		typ:                              typ,
+		targetResource:                   targetResource,
+		schema:                           schema,
+		setPostgresOperationDurationTime: setPostgresOperationDurationTime,
+	}
+}
+
+// Count returns the number of objects in the store.
+func (s *GenericSingleIDStore[T, PT]) Count(ctx context.Context) (int, error) {
+	defer s.setPostgresOperationDurationTime(time.Now(), ops.Count, s.typ)
+
+	var sacQueryFilter *v1.Query
+
+	sacQueryFilter, err := GetReadSACQuery(ctx, s.targetResource)
+	if err != nil {
+		return 0, err
+	}
+
+	return RunCountRequestForSchema(ctx, s.schema, sacQueryFilter, s.db)
+}
+
+// DeleteByQuery removes the objects from the store based on the passed query.
+func (s *GenericSingleIDStore[T, PT]) DeleteByQuery(ctx context.Context, query *v1.Query) error {
```

Comment:
In the store template, `GetByQuery` is only added in the case there is a `search-category` passed to the generator. `DeleteByQuery` is added in any case, but this can lead to problems if the input query has specific constraints (see below).
The problem is when the passed query contains a `MatchFieldQuery` subquery, where access to the OptionsMap is needed. Otherwise, for the `DocID` case which is used for example in `Delete` and `DeleteMany`, this is OK.

Reply:
Good point. I added it here as it's a generic store with all possible functions. The final store should be accessed by an interface thus only subset of this methods will be public.